src/lib/JAX/Concrete.hs

```diff
@@ -0,0 +1,296 @@
+-- Copyright 2023 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+{-# LANGUAGE DuplicateRecordFields #-}
+
+module JAX.Concrete where
+
+import GHC.Generics (Generic (..))
+
+import Data.Aeson
+import Data.Aeson.Types qualified as A
+
+import IRVariants
+import Name
+
+data Primitive =
+    Sin | Add
+  | Scan ScanParams
+  | ConvertElementType ConvertElementTypeParams
+  -- others!
+  deriving (Generic)
+
+data ScanParams = ScanParams
+  { reverse :: Bool
+  , length :: Int
+  , jaxpr :: (ClosedJaxpr VoidS) -- The scan body
+  , num_consts :: Int
+  , num_carry :: Int
+  , linear :: [Bool] -- which arguments are we linear wrt?
+  , unroll :: Int -- how many copies of the body to replicate
+  }
+  deriving (Generic)
+
+data ConvertElementTypeParams = ConvertElementTypeParams
+  { new_dtype :: DType
+  , weak_type :: Bool
+  }
+  deriving (Generic)
+
+data DType = F64 | F32 | I64 | I32 -- others
+  deriving (Generic)
+
+data DimSizeDeBrujin = LitDimSize Int
+  | RefDimSizeInput Int  -- "de Brujin" index but counted from the left of the list
+  | RefDimSizeOutput Int -- same
+  deriving (Generic)
+
+data JArgType = JArrayDeBrujin DType [DimSizeDeBrujin]
+-- On the output, can refer to input binders and preceding things in the output
+  deriving (Generic)
+
+data JFuncType = JFunc [JArgType] JEffects [JArgType]
+-- The ints just count from the beginning of the list of inputs of the jaxpr
+-- being typed by this JEffects datum
+  deriving (Generic)
+
+-- TODO Variable references, polynomials, etc
+data DimSizeName = DimSize Int
+  deriving (Generic)
+
+data JVarType = JArrayName
+  { shape :: [DimSizeName]
+  , dtype :: DType
+  }
+  deriving (Generic)
+
+data JEffects = IO | Read Int | Write Int | Append Int
+  deriving (Generic)
+
+data JAtom (n::S) =
+    JVariable (JVar n)
+  | JLiteral JLit
+           -- DropVar?
+  deriving (Generic)
+
+data JSourceName = JSourceName
+  { name :: Int
+  , count :: Int
+  , suffix :: String
+  }
+  deriving (Eq, Ord, Show)
+
+instance HasNameHint JSourceName where
+  getNameHint JSourceName{suffix=suffix} = getNameHint suffix
+
+data JSourceNameOr (a::E) (n::S) where
+  SourceName :: JSourceName -> JSourceNameOr a 'VoidS
+  InternalName :: JSourceName -> a n -> JSourceNameOr a n
+
+data JVar (n::S) = JVar
+  { sourceName :: JSourceNameOr (Name (AtomNameC SimpIR)) n
```

Comment:
We probably want to give `JVar` its own name color at some point instead of using `SAtomName`. But this is fine for now.

Reply:
Noted.